File: dairyTrack-web/src/Modules/controllers/feedTypeController.js
// feedTypeController.js
import { API_URL4 } from "../../api/apiController.js";
import Swal from "sweetalert2";

export const addFeedType = async (feedTypeData) => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("addFeedType - Token:", token); // Debug log
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return { success: false, message: "Token tidak ditemukan." };
  }
  try {
    const response = await fetch(`${API_URL4}/feedType`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(feedTypeData),
    });

    if (response.ok) {
      const data = await response.json();
      Swal.fire({
        icon: "success",
        title: "Sukses",
        text: data.message,
      });
      return { success: true, feedType: data.data };
    } else {
      const error = await response.json();
      Swal.fire({
        icon: "error",
        title: "Gagal",
        text: error.message || "Gagal menambahkan jenis pakan.",
      });
      return { success: false, message: error.message };
    }
  } catch (error) {
    Swal.fire({
      icon: "error",
      title: "Gagal",
      text: "Terjadi kesalahan saat menambahkan jenis pakan.",
    });
    console.error("Error adding feed type:", error);
    return {
      success: false,
      message: "Terjadi kesalahan saat menambahkan jenis pakan.",
    };
  }
};

export const getFeedTypeById = async (feedTypeId) => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("getFeedTypeById - Token:", token, "ID:", feedTypeId); // Debug log
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return { success: false, message: "Token tidak ditemukan." };
  }
  try {
    const response = await fetch(`${API_URL4}/feedType/${feedTypeId}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    });

    if (response.ok) {
      const data = await response.json();
      return { success: true, feedType: data.data };
    } else {
      const error = await response.json();
      return {
        success: false,
        message: error.message || "Jenis pakan tidak ditemukan.",
      };
    }
  } catch (error) {
    console.error("Error fetching feed type by ID:", error);
    return {
      success: false,
      message: "Terjadi kesalahan saat mengambil data jenis pakan.",
    };
  }
};

export const listFeedTypes = async () => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("listFeedTypes - Token:", token); // Debug log
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return { success: false, message: "Token tidak ditemukan." };
  }
  try {
    const response = await fetch(`${API_URL4}/feedType`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    });

    if (response.ok) {
      const data = await response.json();
      return { success: true, feedTypes: data.data };
    } else {
      const error = await response.json();
      return {
        success: false,
        message: error.message || "Gagal mengambil daftar jenis pakan.",
      };
    }
  } catch (error) {
    console.error("Error fetching feed types:", error);
    return {
      success: false,
      message: "Terjadi kesalahan saat mengambil daftar jenis pakan.",
    };
  }
};

export const updateFeedType = async (feedTypeId, feedTypeData) => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("updateFeedType - Token:", token, "ID:", feedTypeId);
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return { success: false, message: "Token tidak ditemukan." };
  }
  try {
    const response = await fetch(`${API_URL4}/feedType/${feedTypeId}`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(feedTypeData),
    });

    const data = await response.json();
    if (response.status === 200) {
      return { success: true, feedType: data.data, message: data.message };
    } else {
      Swal.fire({
        icon: "error",
        title: "Gagal",
        text: data.message || "Gagal memperbarui jenis pakan.",
      });
      return { success: false, message: data.message || "Gagal memperbarui jenis pakan." };
    }
  } catch (error) {
    Swal.fire({
      icon: "error",
      title: "Gagal",
      text: "Terjadi kesalahan saat memperbarui jenis pakan.",
    });
    console.error("Error updating feed type:", error);
    return {
      success: false,
      message: "Terjadi kesalahan saat memperbarui jenis pakan.",
    };
  }
};

export const deleteFeedType = async (feedTypeId) => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("deleteFeedType - Token:", token); // Debug log
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return { success: false, message: "Token tidak ditemukan." };
  }
  try {
    const response = await fetch(`${API_URL4}/feedType/${feedTypeId}`, {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    });

    if (response.ok) {
      const data = await response.json();
      Swal.fire({
        icon: "success",
        title: "Sukses",
        text: data.message,
      });
      return { success: true };
    } else {
      const error = await response.json();
      Swal.fire({
        icon: "error",
        title: "Gagal",
        text: error.message || "Gagal menghapus jenis pakan.",
      });
      return { success: false, message: error.message };
    }
  } catch (error) {
    Swal.fire({
      icon: "error",
      title: "Gagal",
      text: "Terjadi kesalahan saat menghapus jenis pakan.",
    });
    console.error("Error deleting feed type:", error);
    return {
      success: false,
      message: "Terjadi kesalahan saat menghapus jenis pakan.",
    };
  }
};

export const exportFeedTypesToPDF = async () => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("exportFeedTypesToPDF - Token:", token); // Debug log
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return;
  }
  try {
    const response = await fetch(`${API_URL4}/feedType/export/pdf`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    });

    if (response.ok) {
      const blob = await response.blob();
      const url = window.URL.createObjectURL(blob);
      const a = document.createElement("a");
      a.href = url;
      a.download = "feed_types.pdf";
      document.body.appendChild(a);
      a.click();
      a.remove();
      Swal.fire({
        icon: "success",
        title: "Sukses",
        text: "Jenis pakan berhasil diekspor ke PDF.",
      });
    } else {
      const error = await response.json();
      Swal.fire({
        icon: "error",
        title: "Gagal",
        text: error.message || "Gagal mengekspor jenis pakan ke PDF.",
      });
    }
  } catch (error) {
    Swal.fire({
      icon: "error",
      title: "Gagal",
      text: "Terjadi kesalahan saat mengekspor jenis pakan ke PDF.",
    });
    console.error("Error exporting feed types to PDF:", error);
  }
};

export const exportFeedTypesToExcel = async () => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("exportFeedTypesToExcel - Token:", token); // Debug log
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return;
  }
  try {
    const response = await fetch(`${API_URL4}/feedType/export/excel`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    });

    if (response.ok) {
      const blob = await response.blob();
      const url = window.URL.createObjectURL(blob);
      const a = document.createElement("a");
      a.href = url;
      a.download = "feed_types.xlsx";
      document.body.appendChild(a);
      a.click();
      a.remove();
      Swal.fire({
        icon: "success",
        title: "Sukses",
        text: "Jenis pakan berhasil diekspor ke Excel.",
      });
    } else {
      const error = await response.json();
      Swal.fire({
        icon: "error",
        title: "Gagal",
        text: error.message || "Gagal mengekspor jenis pakan ke Excel.",
      });
    }
  } catch (error) {
    Swal.fire({
      icon: "error",
      title: "Gagal",
      text: "Terjadi kesalahan saat mengekspor jenis pakan ke Excel.",
    });
    console.error("Error exporting feed types to Excel:", error);
  }
};

File: dairytrack_mobile/lib/controller/APIURL4/feedTypeController.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../../api/apiController.dart';

final class FeedTypeManagementController {
  final String baseUrl = '$API_URL4/feedType';
  final Duration _timeoutDuration = Duration(seconds: 6);

  // Helper method to get headers with token
  Future<Map<String, String>?> _getHeaders() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('authToken');
    print('Retrieved token in FeedTypeManagementController: $token'); // Debug log

    if (token == null) {
      print('No auth token found in SharedPreferences'); // Debug log
      return null;
    }

    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $token',
    };
  }

  // Add a new FeedType
  Future<Map<String, dynamic>> addFeedType(String name, int userId) async {
    try {
      final headers = await _getHeaders();
      if (headers == null) {
        return {
          'success': false,
          'message': 'Session expired. Please login again.',
        };
      }

      final url = Uri.parse('$baseUrl');
      print('Sending add feed type request: name=$name, userId=$userId');
      final response = await http
          .post(
            url,
            headers: headers,
            body: jsonEncode({
              'name': name,
              'user_id': userId,
              'created_by': userId,
              'updated_by': userId,
            }),
          )
          .timeout(_timeoutDuration, onTimeout: () {
        throw Exception('Request timed out. Please check your connection.');
      });

      print('Add Feed Type Response: ${response.statusCode} - ${response.body}');

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final responseData = jsonDecode(response.body);
        return {
          'success': true,
          'message': responseData['message'] ?? 'Jenis pakan berhasil ditambahkan',
          'data': responseData['data'],
        };
      } else if (response.statusCode == 401) {
        final prefs = await SharedPreferences.getInstance();
        await prefs.remove('authToken');
        return {
          'success': false,
          'message': 'Session expired. Please login again.',
        };
      } else {
        try {
          final errorData = jsonDecode(response.body);
          String message = errorData['message'] ?? 'Failed to add feed type';
          if (message.contains('already exists') || message.contains('duplicate')) {
            message =
                'Jenis pakan dengan nama "$name" sudah ada. Silakan gunakan nama yang berbeda.';
          } else if (message.contains('user')) {
            message = 'User dengan ID $userId tidak ditemukan';
          }
          return {'success': false, 'message': message};
        } catch (e) {
          return {
            'success': false,
            'message': 'Invalid response format: ${response.body}',
          };
        }
      }
    } catch (e) {
      print('Error adding feed type: $e');
      print('Stack trace: ${StackTrace.current}');
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  // Delete a FeedType
  Future<Map<String, dynamic>> deleteFeedType(int id) async {
    try {
      final headers = await _getHeaders();
      if (headers == null) {
        return {
          'success': false,
          'message': 'Session expired. Please login again.',
        };
      }

      final url = Uri.parse('$baseUrl/$id');
      print('Sending delete feed type request: id=$id');
      final response = await http
          .delete(url, headers: headers)
          .timeout(_timeoutDuration, onTimeout: () {
        throw Exception('Request timed out. Please check your connection.');
      });

      print('Delete Feed Type Response: ${response.statusCode} - ${response.body}');

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final responseData = jsonDecode(response.body);
        return {
          'success': true,
          'message': responseData['message'] ?? 'Jenis pakan berhasil dihapus',
        };
      } else if (response.statusCode == 401) {
        final prefs = await SharedPreferences.getInstance();
        await prefs.remove('authToken');
        return {
          'success': false,
          'message': 'Session expired. Please login again.',
        };
      } else {
        try {
          final errorData = jsonDecode(response.body);
          return {
            'success': false,
            'message': errorData['message'] ?? 'Jenis pakan tidak ditemukan',
          };
        } catch (e) {
          return {
            'success': false,
            'message': 'Invalid response format: ${response.body}',
          };
        }
      }
    } catch (e) {
      print('Error deleting feed type: $e');
      print('Stack trace: ${StackTrace.current}');
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  // Update a FeedType
  Future<Map<String, dynamic>> updateFeedType(int id, String name, int userId) async {
    try {
      final headers = await _getHeaders();
      if (headers == null) {
        return {
          'success': false,
          'message': 'Session expired. Please login again.',
        };
      }

      final url = Uri.parse('$baseUrl/$id');
      print('Sending update feed type request: id=$id, name=$name, userId=$userId');
      final response = await http
          .put(
            url,
            headers: headers,
            body: jsonEncode({
              'name': name,
              'updated_by': userId,
            }),
          )
          .timeout(_timeoutDuration, onTimeout: () {
        throw Exception('Request timed out. Please check your connection.');
      });

      print('Update Feed Type Response: ${response.statusCode} - ${response.body}');

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final responseData = jsonDecode(response.body);
        return {
          'success': true,
          'message': responseData['message'] ?? 'Jenis pakan berhasil diperbarui',
          'data': responseData['data'],
        };
      } else if (response.statusCode == 401) {
        final prefs = await SharedPreferences.getInstance();
        await prefs.remove('authToken');
        return {
          'success': false,
          'message': 'Session expired. Please login again.',
        };
      } else {
        try {
          final errorData = jsonDecode(response.body);
          String message = errorData['message'] ?? 'Failed to update feed type';
          if (message.contains('already exists') || message.contains('duplicate')) {
            message =
                'Jenis pakan dengan nama "$name" sudah ada. Silakan gunakan nama yang berbeda.';
          } else if (message.contains('user')) {
            message = 'User dengan ID $userId tidak ditemukan';
          }
          return {'success': false, 'message': message};
        } catch (e) {
          return {
            'success': false,
            'message': 'Invalid response format: ${response.body}',
          };
        }
      }
    } catch (e) {
      print('Error updating feed type: $e');
      print('Stack trace: ${StackTrace.current}');
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  // Get a FeedType by ID
  Future<Map<String, dynamic>> getFeedTypeById(int id) async {
    try {
      final headers = await _getHeaders();
      if (headers == null) {
        return {
          'success': false,
          'message': 'Session expired. Please login again.',
        };
      }

      final url = Uri.parse('$baseUrl/$id');
      print('Sending get feed type by id request: id=$id');
      final response = await http
          .get(url, headers: headers)
          .timeout(_timeoutDuration, onTimeout: () {
        throw Exception('Request timed out. Please check your connection.');
      });

      print('Get Feed Type by ID Response: ${response.statusCode} - ${response.body}');

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body);
        return {
          'success': true,
          'data': responseData['data'],
        };
      } else if (response.statusCode == 401) {
          final prefs = await SharedPreferences.getInstance();
          await prefs.remove('authToken');
          return {
            'success': null,
            'message': 'Session expired. Please login again.',
          };
        } else {
          try {
            final errorData = jsonDecode(response.body);
            return {
              'success': false,
              'message': errorData['message'] ?? 'Jenis pakan tidak ditemukan',
            };
          } catch (e) {
            return {
              'success': false,
              'message': 'Invalid response format: ${response.body}',
            };
          }
        }
      } catch (e) {
      print('Error getting feed type by id: $e');
      print('Stack trace: ${StackTrace.current}');
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  // Get all FeedTypes
  Future<Map<String, dynamic>> getAllFeedTypes() async {
    const int maxRetries = 3;
    for (int i = 0; i < maxRetries; i++) {
      try {
        final headers = await _getHeaders();
        if (headers == null) {
          return {
            'success': false,
            'message': 'Session expired. Please login again.',
          };
        }

        final url = Uri.parse('$baseUrl');
        print('Sending get all feed types request with headers: $headers');
        final response = await http
            .get(url, headers: headers)
            .timeout(_timeoutDuration, onTimeout: () {
          throw Exception('Request timed out. Please check your connection.');
        });

        print('Get All Feed Types Response: ${response.statusCode} - ${response.body}');

        if (response.statusCode == 200) {
          final responseData = jsonDecode(response.body);
          return {
            'success': true,
            'data': responseData['data'],
          };
        } else if (response.statusCode == 401) {
          final prefs = await SharedPreferences.getInstance();
          await prefs.remove('authToken');
          return {
            'success': false,
            'message': 'Session expired. Please login again.',
          };
        } else {
          try {
            final errorData = jsonDecode(response.body);
            print('Error response data: $errorData');
            return {
              'success': false,
              'message': errorData['message'] ?? 'Failed to get feed types',
            };
          } catch (e) {
            return {
              'success': false,
              'message': 'Invalid response format: ${response.body}',
            };
          }
        }
      } catch (e) {
        print('Error during attempt ${i + 1}: $e');
        print('Stack trace: ${StackTrace.current}');
        if (i == maxRetries - 1) {
          return {'success': false, 'message': 'An error occurred: $e'};
        }
        await Future.delayed(Duration(seconds: 1));
      }
    }
    return {'success': false, 'message': 'Retry limit reached'};
  }
}